File: backend/node/src/product/entities/product.ts
import { Entity, Column, PrimaryGeneratedColumn, Index } from 'typeorm';

/*
Storing CurrencyConversions as pre-calculated columns for perfromance reasons
Calculated columns would be slow for large datasets
We meet the current requirements, any further changes may require refactorings
*/
@Entity()
export class Product {
  @PrimaryGeneratedColumn()
  Id: number;

  @Index()
  @Column({ type: 'varchar', length: 255 })
  Name: string;

  @Index()
  @Column({ type: 'decimal', precision: 10, scale: 2 })
  Price: number;

  @Index()
  @Column({ type: 'date' })
  Expiration: Date;

  @Index()
  @Column({ type: 'date' })
  CreatedAt: Date;

  // Brazilian Reais
  @Column({ type: 'decimal', precision: 10, scale: 2, nullable: true })
  BRL: number | null;

  // Euro
  @Column({ type: 'decimal', precision: 10, scale: 2, nullable: true })
  EUR: number | null;

  // Canadian dollar
  @Column({ type: 'decimal', precision: 10, scale: 2, nullable: true })
  CAD: number | null;

  // Mexican Pesos
  @Column({ type: 'decimal', precision: 10, scale: 2, nullable: true })
  MXN: number | null;

  // Libras
  @Column({ type: 'decimal', precision: 10, scale: 2, nullable: true })
  GBP: number | null;
}

File: backend/_others_/cs_dapper/Models/Product.cs
namespace cssharp.Models
{
  public class Product
  {
    public int Id { get; set; }

    public string Name { get; set; } = "";

    public decimal Price { get; set; }

    public DateTime Expiration { get; set; }

    public DateTime CreatedAt { get; set; }

    // Brazilian Reais  
    public decimal BRL { get; set; }

    // Euro
    public decimal EUR { get; set; }

    // Canadian dollar
    public decimal CAD { get; set; }

    // Mexican Pesos
    public decimal MXN { get; set; }

    // Libras
    public decimal GBP { get; set; }
  }
}
